Update scriptcanvas nodeable node doc (#1782)
o3de/o3de.org: content/docs/user-guide/scripting/script-canvas/programmer-guide/custom-nodes/custom-nodeable-nodes.md

   AUTOGEN_RULES
       *.ScriptCanvasNodeable.xml,ScriptCanvasNodeable_Header.jinja,$path/$fileprefix.generated.h
       *.ScriptCanvasNodeable.xml,ScriptCanvasNodeable_Source.jinja,$path/$fileprefix.generated.cpp

>        *.ScriptCanvasNodeable.xml,ScriptCanvasNodeableRegistry_Header.jinja,AutoGenNodeableRegistry.generated.h

Author: Actually we don't expect customer to use ScriptCanvasGrammar, right now ScriptCanvasGrammar is all for our internal usage and it is not a nodeable node.

Plus StartingPointInput Gem ScriptCanvasGrammar usage is even for deperecation. (We don't have any other Gem as a good reference)

        , protected InputEventNotificationBus::Handler
    {
        SCRIPTCANVAS_NODE(InputHandlerNodeable)                                                      // (2)
        ...
    };
}
```

## Step 4: Add source files to CMake {#add-source-files-to-cmake}

Add the XML and class source files to one of Gem's .cmake files.

For example, in `InputHandlerNodeable` we must add the following lines:

File: [InputHandlerNodeable.h](https://github.com/o3de/o3de/blob/development/Gems/StartingPointInput/Code/Source/InputHandlerNodeable.h)
```cmake
set(FILES
    ...
    Source/InputHandlerNodeable.h
    Source/InputHandlerNodeable.cpp
    Source/InputHandlerNodeable.ScriptCanvasNodeable.xml
    ...
)
```

## Step 5: Reflect the new node {#reflect-the-new-node}
{{< note >}}
This step is only required once for the first time nodeable node creation.
{{< /note >}}

The final step is to register and reflect the new node. To do this, you need to modify Gem [Module](/docs/user-guide/programming/gems/overview/) and [System Component](/docs/user-guide/programming/components/system-components/). Use the **StartingPointInput** Gem from the O3DE source as a reference:

> 

Author: REGISTER_SCRIPTCANVAS_AUTOGEN_FUNCTION is for custom function node which is covered by a different topic

REGISTER_SCRIPTCANVAS_AUTOGEN_GRAMMAR is for internal usage right now only, and some use case is related to deprecation node.

1. Include auto-generated registry header file, and invoke `REGISTER_SCRIPTCANVAS_AUTOGEN_NODEABLE` with sanitized Gem target name. 
    {{< note >}}
    Auto-generated registry header file should be the same name declared under `AUTOGEN_RULES` in your Gem's `Code/CMakeLists.txt`. In **StartingPointInput** example, it is `AutoGenNodeableRegistry.generated.h`.
    {{< /note >}}
    {{< note >}}
    Sanitized Gem target name should contain letters and numbers only. In **StartingPointInput** example, it is `StartingPointInputStatic` which refers to `StartingPointInput.Static` target.
    {{< /note >}}
   
    For example, in `StartingPointInputGem.cpp`
  
    ```cpp
    #include <AutoGenFunctionRegistry.generated.h>
    

>     REGISTER_SCRIPTCANVAS_AUTOGEN_FUNCTION(MyGemStatic)

Author: good catch, copy paste mistake, for nodeable usage, we just need Nodeable related header and macros

    ...
    ```

1. Reflect and initialize auto-generated registry through Gem system component Reflect function.

    For example, in `StartingPointInputGem.cpp`

    ```cpp
    void StartingPointInputSystemComponent::Reflect(AZ::ReflectContext* context)
    {
        REFLECT_SCRIPTCANVAS_AUTOGEN(StartingPointInputStatic, context);
        ...
    }
    ```
   
    ```cpp
    void StartingPointInputSystemComponent::Init()
    {
        INIT_SCRIPTCANVAS_AUTOGEN(StartingPointInputStatic);
    }   
    ```

1. Insert auto-generated nodeable descriptors in Gem module.

    For example, in `StartingPointInputGem.cpp`
    ```cpp
    StartingPointInputModule() : AZ::Module()
    {
        ...
        AZStd::vector<AZ::ComponentDescriptor*> autogenDescriptors(GET_SCRIPTCANVAS_AUTOGEN_COMPONENT_DESCRIPTORS(StartingPointInputStatic));
        m_descriptors.insert(m_descriptors.end(), autogenDescriptors.begin(), autogenDescriptors.end());
    }
    ```

## Advanced ScriptCanvasNodeable.xml Usage
This topic explores additional features we support in nodeable XML file.

### Base and derived nodeable node

> If you have basic shared logic across multiple nodeable node, we do support create base and derived nodeable node.
> We'll use the following XML, copied from the O3DE source for the **Time Delay** node.

Reviewer: ```suggestion
If you have shared logic across multiple nodeable nodes, you can create a base node and multiple derived nodes.

The following example uses the O3DE source for the **Time Delay** node:
```

```xml
<ScriptCanvas Include="Include/ScriptCanvas/Libraries/Time/TimeDelayNodeable.h" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <Class Name="TimeDelayNodeable"
        QualifiedName="Nodeables::Time::TimeDelayNodeable"
        PreferredClassName="Time Delay"
        Base="Nodeables::Time::BaseTimer"    # Declare base class as Nodeables::Time::BaseTimer
        Category="Timing"
        Namespace="ScriptCanvas"
        Description="Delays all incoming execution for the specified number of ticks">

        <Input Name="Start" Description="">
            <Parameter Name="Delay" Type="Data::NumberType" DefaultValue="0.0" Description="The amount of time to delay before the Done is signalled."/>
        </Input>

        <Output Name="Done" Description="Signaled after waiting for the specified amount of times."/>

        <PropertyInterface Property="m_timeUnitsInterface" Name="Units" Type="Input" Description="Units to represent the time in."/>
    </Class>
</ScriptCanvas>
```

> The `TimeDelayNodeable` class itself implements a base class, called `BaseTimer`. In the following XML, you can see the base class defines shared properties

Reviewer: ```suggestion
The `TimeDelayNodeable` class implements a base class, called `BaseTimer`. In the following base class XML, you can see that the base class defines the shared properties, "Units" and "TickOrder":

File: [BaseTimer.ScriptCanvasNodeable.xml](https://github.com/o3de/o3de/blob/development/Gems/ScriptCanvas/Code/Include/ScriptCanvas/Internal/Nodeables/BaseTimer.ScriptCanvasNodeable.xml)

```

        Uuid="{64814C82-DAE5-9B04-B375-5E47D51ECD26}"
        BaseClass="True"    # Declare this nodeable as a base class, so it won't be reflected as a node
        Category="Timing"
        Description="Provides a basic interaciton layer for all time based nodes for users(handles swapping between ticks and seconds).">

        <Property Name="m_timeUnits" Type="int" DefaultValue="0" Serialize="true">
            <PropertyData Name="Units"
                Description="Units to represent the time in."
                Serialize="true"
                UIHandler="AZ::Edit::UIHandlers::ComboBox">
                <EditAttribute Key="AZ::Edit::Attributes::GenericValueList" Value="&amp;BaseTimer::GetTimeUnitList"/>
                <EditAttribute Key="AZ::Edit::Attributes::PostChangeNotify" Value="&amp;BaseTimer::OnTimeUnitsChanged"/>
            </PropertyData>
        </Property>

        <Property Name="m_tickOrder" Type="int" DefaultValue="static_cast&lt;int&gt;(AZ::TICK_DEFAULT)" Serialize="true">
            <PropertyData Name="TickOrder" Description="When the tick for this time update should be handled."/>
        </Property>
    </Class>
</ScriptCanvas>
```

{{< note >}}

> For further node name, tooltip and category customization, please refer to [Text Replacement](/docs/user-guide/scripting/script-canvas/editor-reference/text-replacement/)

Reviewer: ```suggestion
For further node name, tooltip, and category customization, please refer to [Text Replacement](/docs/user-guide/scripting/script-canvas/editor-reference/text-replacement/).
```
